src/torchmetrics/utilities/data.py

```diff
@@ -191,11 +191,12 @@ def _bincount(x: Tensor, minlength: Optional[int] = None) -> Tensor:
     """
     if minlength is None:
         minlength = len(torch.unique(x))
+
     if torch.are_deterministic_algorithms_enabled() or _XLA_AVAILABLE or _TORCH_GREATER_EQUAL_1_12 and x.is_mps:
-        output = torch.zeros(minlength, device=x.device, dtype=torch.long)
-        for i in range(minlength):
-            output[i] = (x == i).sum()
-        return output
+        size = len(x)
+        mesh = torch.meshgrid(torch.arange(size), torch.arange(minlength), indexing="ij")[1].to(x.device)
```

Comment:
```suggestion
        mesh = torch.arange(minlength, device=x.device).repeat(size, 1)
```
some small experimentation tells me this should be faster
![image](https://github.com/Lightning-AI/torchmetrics/assets/24896311/9ab5039e-9b29-4d72-9a93-dd61d17cefbd)